File: src/movieclaw_api/schemas/jobs.py
"""持久化后台任务的公共 API 契约。"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import Field

from movieclaw_api.schemas.base import BaseModel
from movieclaw_db.models import JobStatus


class JobProgressView(BaseModel):
    mode: str = Field(description="determinate / indeterminate / waiting / paused")
    phase: str = Field(description="当前领域阶段的稳定标识")
    message: str = Field(description="面向用户的即时进展")
    current: int | None = Field(default=None, description="当前完成量")
    total: int | None = Field(default=None, description="总量；未知时为空")
    percent: float | None = Field(default=None, description="真实可计算时才返回百分比")
    phase_index: int | None = Field(default=None, description="当前阶段序号")
    phase_count: int | None = Field(default=None, description="总阶段数")
    details: dict[str, Any] = Field(default_factory=dict, description="领域进度明细")


class JobResourceView(BaseModel):
    resource_type: str
    resource_id: str
    relation: str


class JobView(BaseModel):
    id: str
    job_type: str
    subject: str | None
    definition_version: int
    handler_revision: str
    prompt_revision: str | None
    provider_ref: str | None
    status: JobStatus
    input_data: dict[str, Any] = Field(description="已确认的任务输入（永不包含密钥）")
    progress: JobProgressView
    result: dict[str, Any] | None
    error: dict[str, Any] | None
    usage: dict[str, Any]
    resources: list[JobResourceView] = Field(default_factory=list)
    origin: str
    actor_kind: str | None
    actor_name: str | None
    parent_job_id: str | None
    root_job_id: str | None
    retry_of_job_id: str | None
    attempt: int
    max_attempts: int
    revision: int
    cancel_requested_at: datetime | None
    cancel_requested_by: str | None = Field(
        default=None,
        description="取消发起方；system: 前缀代表系统自动收口，前端据此隐藏「重新执行」",
    )
    dismissed_at: datetime | None = Field(
        default=None,
        description="用户忽略的时间；非空表示不再计入「需要处理」，但任务本身仍是失败",
    )
    dismissed_by: str | None = Field(default=None, description="忽略操作者")
    created_at: datetime
    updated_at: datetime
    started_at: datetime | None
    finished_at: datetime | None


class JobListView(BaseModel):
    items: list[JobView]


class JobWaitView(BaseModel):
    changed: bool
    job: JobView


class JobCancelView(BaseModel):
    cancelled: bool
    job: JobView


class JobRetryView(BaseModel):
    created: bool
    job: JobView


class JobDismissRequest(BaseModel):
    mute_source: bool = Field(
        default=False,
        description=(
            "同时静音自动来源：不再自动为该对象重建同类任务。"
            "只对有自动来源的任务类型有效（当前为字幕自动生成）；手动触发不受影响"
        ),
    )


class JobDismissView(BaseModel):
    dismissed: bool = Field(description="本次是否真的改变了状态；重复忽略为 false")
    muted: bool = Field(default=False, description="是否落了自动来源静音记录")
    job: JobView


class JobDismissAllRequest(BaseModel):
    job_type: str | None = Field(default=None, description="只忽略该类型；留空表示全部")


class JobDismissAllView(BaseModel):
    dismissed: int = Field(description="本次忽略的任务条数")
    jobs: list[JobView]


class JobEventView(BaseModel):
    id: int
    job_id: str
    revision: int
    event_type: str
    payload: dict[str, Any]
    created_at: datetime


class JobEventListView(BaseModel):
    items: list[JobEventView]


class JobWorkerHealthView(BaseModel):
    running: bool
    owner: str | None
    active_workers: int
    counts: dict[str, int]
    oldest_queued_seconds: int | None

File: apps/web/lib/api/jobs.ts
import { request } from "@/lib/http";

interface ApiEnvelope<T> {
  success: boolean;
  code: string;
  message: string;
  data: T;
}

export type JobStatus =
  | "queued"
  | "running"
  | "retry_wait"
  | "cancelling"
  | "waiting"
  | "blocked"
  | "succeeded"
  | "failed"
  | "cancelled";

export interface JobProgress {
  mode: "determinate" | "indeterminate" | "waiting" | "paused";
  phase: string;
  message: string;
  current: number | null;
  total: number | null;
  percent: number | null;
  phase_index: number | null;
  phase_count: number | null;
  details: Record<string, unknown>;
}

export interface JobResource {
  resource_type: string;
  resource_id: string;
  relation: string;
}

export interface JobView {
  id: string;
  job_type: string;
  subject: string | null;
  definition_version: number;
  handler_revision: string;
  prompt_revision: string | null;
  provider_ref: string | null;
  status: JobStatus;
  input_data: Record<string, unknown>;
  progress: JobProgress;
  result: Record<string, unknown> | null;
  error: {
    code?: string;
    message?: string;
    details?: Record<string, unknown>;
    actions?: Array<{ type: string; label: string; target?: string }>;
  } | null;
  usage: Record<string, unknown>;
  resources: JobResource[];
  origin: string;
  actor_kind: string | null;
  actor_name: string | null;
  parent_job_id: string | null;
  root_job_id: string | null;
  retry_of_job_id: string | null;
  attempt: number;
  max_attempts: number;
  revision: number;
  cancel_requested_at: string | null;
  /** 取消发起方；`system:` 前缀代表系统自动收口，重跑没有意义，界面不给「重新执行」。 */
  cancel_requested_by: string | null;
  /** 用户忽略的时间。非空 = 不再计入「需要处理」；任务本身仍然是失败，日志与重试都还在。 */
  dismissed_at: string | null;
  dismissed_by: string | null;
  created_at: string;
  updated_at: string;
  started_at: string | null;
  finished_at: string | null;
}

/** 系统自动收口的取消（如整树入库完成后清理被取代的分批任务）：重跑没有意义，不提供「重新执行」。 */
export function isSystemCancelled(job: JobView): boolean {
  return job.status === "cancelled" && (job.cancel_requested_by?.startsWith("system:") ?? false);
}

export const ACTIVE_JOB_STATUSES = new Set<JobStatus>([
  "queued",
  "running",
  "retry_wait",
  "cancelling",
  "waiting",
  "blocked",
]);

export async function listJobs(options: {
  activeOnly?: boolean;
  limit?: number;
} = {}): Promise<JobView[]> {
  const query = new URLSearchParams({ limit: String(options.limit ?? 100) });
  if (options.activeOnly) query.set("active_only", "true");
  const response = await request<ApiEnvelope<{ items: JobView[] }>>(`/jobs?${query}`);
  return response.data.items;
}

export async function cancelJob(jobId: string): Promise<JobView> {
  const response = await request<ApiEnvelope<{ cancelled: boolean; job: JobView }>>(
    `/jobs/${jobId}/cancel`,
    { method: "POST" },
  );
  return response.data.job;
}

export async function retryJob(jobId: string): Promise<JobView> {
  const response = await request<ApiEnvelope<{ created: boolean; job: JobView }>>(
    `/jobs/${jobId}/retry`,
    { method: "POST" },
  );
  return response.data.job;
}

/**
 * 忽略一条失败任务。
 *
 * `muteSource` 决定这次忽略有没有分量：不勾，下次自动扫描还会为同一个对象
 * 重建同样的任务、再失败一次（issue #221）；勾上则连自动来源一起静音，
 * 手动触发不受影响。
 */
export async function dismissJob(jobId: string, muteSource = false): Promise<JobView> {
  const response = await request<ApiEnvelope<{ dismissed: boolean; muted: boolean; job: JobView }>>(
    `/jobs/${jobId}/dismiss`,
    { method: "POST", body: JSON.stringify({ mute_source: muteSource }) },
  );
  return response.data.job;
}

export async function undismissJob(jobId: string): Promise<JobView> {
  const response = await request<ApiEnvelope<{ dismissed: boolean; muted: boolean; job: JobView }>>(
    `/jobs/${jobId}/undismiss`,
    { method: "POST" },
  );
  return response.data.job;
}

/** 一次收口眼前这批失败任务；服务端单事务完成，不在前端循环发请求。 */
export async function dismissAllFailedJobs(): Promise<JobView[]> {
  const response = await request<ApiEnvelope<{ dismissed: number; jobs: JobView[] }>>(
    "/jobs/dismiss-all",
    { method: "POST", body: JSON.stringify({}) },
  );
  return response.data.jobs;
}
